### DISCUSSION TITLE
Ideas for additional checks

### DISCUSSION THREAD

**Discussion by klauer** [pcdshub/atef#99] [Ideas]
## Question for discussion

What additional checks could be useful that don't fit into the mold of ophyd Device / EPICS PVs value comparisons?

**Comment by klauer** [comment]
Network layer checks (ping?)

**Comment by klauer** [comment]
EPICS IOC status checks?

* These are pretty facility-specific, so could this just be "run external script and verify it returns success"?
* If the PVs don't connect in other configurations, that also could mean the IOC is down

**Comment by ZLLentz** [comment]
We'd have to make a completely separate branch to the configuration for things like this, e.g. "NetworkConfig" or "IOCConfig" that each have their own flavors of checks. It should be doable, albeit maybe some work to fit into the overall frame here with the goal of having uniform/comprehensive test reporting.

**Reply by klauer** [reply]
I had somewhat anticipated free-form checks like this to be part of the guided/active checkout procedures. They could be custom code there and implemented with some atef/nabs-provided framework/utilities or just pre-written plans.

Given the request, I'm rethinking that approach as I can also see this sort of thing belonging in the passive check portion. I'm only a bit worried about the proliferation of these checks and don't want to see a thousand top-level "Configuration" types for each. How about having these wrapped into a "service/status check" of sorts where you get a drop-down list that we have the room to easily add onto:
* Ping (parameters: hosts)
* Check IOCs via IOC manager (parameters: ioc names)
* (Other stuff that gets suggested here)

Or in dataclass-like code: `ServiceStatusCheck(check="ping", parameters={"hosts": ["host1", "host2"]})`

**Reply by ZLLentz** [reply]
Grouping them all under one banner makes a lot more sense than my initial implementation suggestion.

**Comment by christina-pino** [comment]
Another thing that Andrew and I are beginning to discuss is comparing two PVs. In this case we are thinking of comparing <PV Base Name>.VAL vs <PV Base Name>.HIHI or another similar attribute field. Currently Andrew is manually entering them in. For now it's been a good exercise to get accustomed to atef, but in the long run it is going to be a pain to fix them each time the limits change

**Reply by ZLLentz** [reply]
Some of these I wonder if there should be shortcuts for e.g. "verify value is within alarm range"- or if there should be a simple "is this alarming" check? The reporting tool would surely be able to see "hey this is alarming, let me check why and tell you why in the report" without having to manually set that up in a specific way.

In general comparing a PV to an arbitrary PV makes sense- for example, there's a lot of vacuum checks you could do that involve comparing gauges on opposite sides of a valve. I could imagine making basic syntax for:
- compare against another ophyd signal name (put the name of the signal in the value field)
- compare against a full PV (put the full PV name in the value field, maybe steal the pydm ca:// syntax)
- compare against a PV relative to this PV name

**Reply by klauer** [reply]
For the alarms in general, I think the comparisons may need revisiting to incorporate alarm status checks by default (or optionally). We worked around this originally because ophyd doesn't really offer what we need to do it easily.

I like the other ideas @ZLLentz mentions above - especially with the ability to choose PVs versus ophyd `device.component` naming, too.

nalms integration could also come down the line, too.
